packages/node/src/plugins/autostaker/payoutProportionalStrategy.ts

```diff
@@ -0,0 +1,177 @@
+                // choose different sponsorships. Using hash of some operator-specific ID + sponsorshipId
+                // to determine the order. Here we use operatorContractAddress, but it could also
+                // be e.g. the nodeId.
+                const buffer = crypto.createHash('md5').update(operatorContractAddress + id).digest()
+                return buffer.readInt32LE(0)
+            }
+        )
+    ].slice(0, count)        
+}
+
+/*
+ * Calculate the target stakes for each sponsorship:
+ * - for selected sponsorships the stake is minimum stake plus payout-proportional allocation
+ * - for expired sponsorships the stake is zero
+ */
+const getTargetStakes = (
+    stakes: Map<SponsorshipID, WeiAmount>,
+    stakeableSponsorships: Map<SponsorshipID, SponsorshipConfig>,
+    unstakedAmount: WeiAmount,
+    minStakePerSponsorship: WeiAmount,
+    maxSponsorshipCount: number | undefined,
+    operatorContractAddress: string
+): Map<SponsorshipID, WeiAmount> => {
+    const totalStakeableAmount = sum([...stakes.values()]) + unstakedAmount
+    const selectedSponsorships = getSelectedSponsorships(
+        stakes,
+        stakeableSponsorships,
+        totalStakeableAmount,
+        minStakePerSponsorship,
+        maxSponsorshipCount,
+        operatorContractAddress
+    )
+    const minStakePerSponsorshipSum = BigInt(selectedSponsorships.length) * minStakePerSponsorship
+    const payoutProportionalAmount = totalStakeableAmount - minStakePerSponsorshipSum
+    const payoutPerSecSum = sum(selectedSponsorships.map((id) => stakeableSponsorships.get(id)!.payoutPerSec))
+    const targetsForSelected: TargetStake[] = selectedSponsorships.map((id) => [
+        id,
+        minStakePerSponsorship + payoutProportionalAmount * stakeableSponsorships.get(id)!.payoutPerSec / payoutPerSecSum
+    ])
+    const targetsForExpired: TargetStake[] = getExpiredSponsorships(stakes, stakeableSponsorships).map((id) => [
+        id,
+        0n
+    ]) 
+    return new Map([...targetsForSelected, ...targetsForExpired])
+}
+
+export const adjustStakes: AdjustStakesFn = ({
+    operatorState,
+    operatorConfig,
+    stakeableSponsorships,
+    environmentConfig
+}): Action[] => {
+
+    const targetStakes = getTargetStakes(
+        operatorState.stakes,
+        stakeableSponsorships,
+        operatorState.unstakedAmount,
+        environmentConfig.minStakePerSponsorship,
+        operatorConfig.maxSponsorshipCount,
+        operatorConfig.operatorContractAddress
+    )
+
+    const adjustments = [...targetStakes.keys()]
+        .map((sponsorshipId) => ({ 
+            sponsorshipId,
+            difference: targetStakes.get(sponsorshipId)! - (operatorState.stakes.get(sponsorshipId) ?? 0n)
+        }))
+        .filter(({ difference: difference }) => difference !== 0n)
+
+    // fix rounding errors by forcing the net staking to equal unstakedAmount: adjust the largest staking
```

Comment:
I'd do the "tooSmallAdjustments" before this "force net staking to equal unstakedAmount", because there are two points to this forcing:
* that we never run out of tokens
* that all tokens are put to work

I see in the tooSmallAdjustments a safety to never adjust to remove more unstaking than staking, that's good and keeps the token balance on the correct side. Just then the comment "forcing the net staking to equal unstakedAmount" isn't accurate for the end-result.

I think it would make sense to just first filter out those too small adjustments, then do the forcing; that way, net staking will be same as unstakedAmount (i.e. all tokens are put to work) but also we shouldn't be able to run out of tokens.